content/thin-provisioning-redesign.md

```diff
@@ -0,0 +1,137 @@
+a few different angles.
+
+Earlier filesystem extension
+----------------------------
+Stratis used to wait until several gigabytes were left to extend the filesystem. If Stratis didn't
+resize the filesystem quickly enough, the filesystem would run out of space before the extension
+could complete. While this would eventually resolve itself once the filesystem was extended,
+it would cause some unnecessary IO errors. We now extend the filesystem at 50% usage
+to ensure that users always have a large buffer of free space available for even very IO-heavy
+usage patterns.
+
+Parallelized filesystem extension operations
+--------------------------------------------
+Stratis could previously only iterate sequentially through pools. Now stratisd can handle filesystem
+extension on two separate pools in parallel, reducing the latency between the point where
+high usage is detected and the extension operation being performed.
+
+Periodic checks for filesystem usage
+------------------------------------
+Checking filesystem usage used to be a devicemapper event-dependent operation. This led to
+some problems around filesystem extension. A devicemapper event would be generated periodically
+as the filesystem filled up, but if the filesystem failed to extend a few times,
+devicemapper events would no longer be generated once the pool filled up and users would be
+left with a filesystem that couldn't be extended. We've removed our dependency on devicemapper
+events for filesystem operations and use devicemapper events for pool handling exclusively.
+In place, we run periodic checks in the background on filesystems to ensure that even if
+filesystem extension fails multiple times, once the filesystem is ready to be extended,
+stratisd can perform the operation the background, never leaving users in a state where their
+filesystem can't be extended.
+
+Migration and backwards compatibility
+=====================================
+There are two types of changes that require migrations from older versions of stratisd, 
+metadata changes and allocation scheme changes.
+
+Metadata changes
+----------------
+The changes we made required some schema changes in our MDA, the metadata region outside of
+the superblock that records longer form JSON about the specifics of the pool topology. The
+migration should be invisible to the user and will be performed the first time the new
+version of stratisd detects legacy pools. The migration adds some additional devicemapper
+information, information about filesystem limits on a pool, and other bookkeeping information.
+
+Allocation scheme changes
+-------------------------
+As mentioned above, the previous metadata allocation scheme was less precise and allocated
+a larger segment for metadata space than was necessary for the amount of data space present.
+Migration for old pools mainly means that new versions of stratisd will detect that the
```

Comment:
Maybe something like "The most likely consequence of migrating a pool from the old allocation scheme to the new..."